src/libraries/System.Private.CoreLib/src/System/Random.Xoshiro128StarStarImpl.cs

```diff
@@ -90,46 +95,46 @@ public override int Next()
 
             public override int Next(int maxValue)
             {
-                if (maxValue > 1)
+                Debug.Assert(maxValue >= 0);
+
+                ulong randomProduct = (ulong)(uint)maxValue * NextUInt32();
+                uint lowPart = (uint)randomProduct;
+
+                if (lowPart < (uint)maxValue)
                 {
-                    // Narrow down to the smallest range [0, 2^bits] that contains maxValue.
-                    // Then repeatedly generate a value in that outer range until we get one within the inner range.
-                    int bits = BitOperations.Log2Ceiling((uint)maxValue);
-                    while (true)
+                    uint remainder = ((uint)0 - (uint)maxValue) % (uint)maxValue;
+
+                    while (lowPart < remainder)
                     {
-                        uint result = NextUInt32() >> (sizeof(uint) * 8 - bits);
-                        if (result < (uint)maxValue)
-                        {
-                            return (int)result;
-                        }
+                        randomProduct = (ulong)(uint)maxValue * NextUInt32();
+                        lowPart = (uint)randomProduct;
                     }
                 }
 
-                Debug.Assert(maxValue == 0 || maxValue == 1);
-                return 0;
+                return (int)(randomProduct >> (sizeof(uint) * 8));
             }
 
             public override int Next(int minValue, int maxValue)
             {
+                Debug.Assert(minValue <= maxValue);
```

Comment:
```suggestion
                Debug.Assert(minValue  >= 0 && minValue <= maxValue);
```
?

Reply:
As far as I understand, minValue can be negative.
See also the tests/examples in Xoshiro_AlgorithmBehavesAsExpected (eg. https://github.com/dotnet/runtime/blob/f7633f498a8be34bee739b240a0aa9ae6a660cd9/src/libraries/System.Runtime.Extensions/tests/System/Random.cs#L622).

Reply:
I don't see anything that says that negative values aren't allowed when both bounds are given. And the old implementations seem to work fine with negative bounds.

Reply:
As long as we have tests then, great.

Reply:
I see some basic tests for it.